README.md
# Trippin Tacos
[![pipeline status](https://gitlab.stud.idi.ntnu.no/programvareutvikling-v19/gruppe-30/badges/master/pipeline.svg)](https://gitlab.stud.idi.ntnu.no/programvareutvikling-v19/gruppe-30/commits/master)
[![Vue Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://eslint.vuejs.org/rules/)

![Trippin Tacos](/static/logo-long.png "Trippin Tacos")


#### Et restaurantsystem for Trippin Tacos
Trippin Tacos prosjektet er en nettside for et system der man effektivt kan reservere bord, holde oversikt og se statistikk.

Link til prosjektet: https://pu30-5b0f9.firebaseapp.com


## Installasjon/Kommandoer
I terminalen på et vilkårlig operativsystem, "clone" prosjektet lokalt. 
> `$ git clone https://gitlab.stud.idi.ntnu.no/programvareutvikling-v19/gruppe-30.git`

> `$ cd gruppe-30`

Hvis du ikke har node installert må du installere dette. Det kan du gjøre [her](https://nodejs.org/en/)

Installer avhengigheter: 
> `$ npm install`


Starte med utvikling inkludert hot reload: 
> `$ npm run dev`

Naviger til localhost:'portnr her' i nettleseren din for å se nettsiden. Standardport er 3000.

**Man trenger kun disse stegene for å kunne utvikle prosjektet lokalt.
Videre har man flere kommandoer:**


For å generere statisk prosjekt (for statisk utvikling): 
> `$ npm generate`

For å deploye hosting og functions til nettsiden: 
  * `$ firebase deploy`
  * Kun hosting: `$ firebase deploy --only hosting`
  * Kun cloud functions: `$ firebase deploy --only functions`
  * Teste cloud functions lokalt med f.eks Postman: `$ firebase serve --only functions`
  
  
## Kodestil
Vi bruker streng ESLint for å sørge for at all kode følger den samme konvensjonen. Konvensjonen som blir benyttet er
Vue sine anbefalte konvensjoner (A, B, C og ukategorisert) https://eslint.vuejs.org/rules/.

## Kodeeksempler

#### Spørringer til databasen
```JavaScript
// Eksempel som viser hvordan man kan gjøre spørringer til databasen. Denne koden oppretter en gjestebruker
createGuestUser (payload) {
    const userObject = {
      firstName: payload.firstName || '',
      lastName: payload.lastName || '',
      email: payload.email || '',
      mobile: payload.mobile || ''
    }
    return fs.collection('guestUsers').add(userObject)
      .then(guestID => {
        userObject.guestID = guestID.id
        return fs.collection('guestUsers').doc(userObject.guestID + '').set(userObject)
          .then(() => {
            console.log(guestID)
            return guestID.id
          })
      })
      .catch(error => {
        console.log('Klarte ikke å lage gjestebruker')
        console.log(error)
      })
  },
```
#### Bruke storen
Dette eksempelet viser de ulike delene som går med på å bruke storen til å logge inn en bruker.

**pages/login.vue:**
```JavaScript
// Kaller på storen for å logge oss inn
    onSignin () {
    // Man kaller på async funksjonene i storen, kalt actions, slik som dette:
      this.$store.dispatch('signUserIn', { email: this.email, password: this.password })
    }

```

**store/index.js:**
```JavaScript
// Kaller på storen for å logge oss inn
    /* signUserIn brukes av login for å logge inn brukeren. Først autentiserers den hos google, før vi henter dataene
  * vi har lagret på brukeren */
  signUserIn ({ commit }, payload) {
    commit('setLoading', true)
    commit('clearError')
    firebase.auth().signInWithEmailAndPassword(payload.email, payload.password)
      .then(user => {
        commit('setLoading', false)
        firebase.firestore().collection('users').doc(user.user.uid).get()
          .then(user => {
            user = user.data()
            commit('setUser', user)
          })
      })
      .catch(error => {
        commit('setLoading', false)
        commit('setError', error)
        console.log(error)
      })
  },

```
Etter at denne metoden har kjørt, ser vi at den har kalt på commit flere ganger. Commit er metoder i storen som er sync og som endrer staten (mutations):
**store/index.js:**
```JavaScript
  // Setter brukeren og om han er admin og/eller employee
  setUser (state, payload) {
    state.user = payload
    state.admin = payload.admin
    state.employee = payload.employee
    state.customer = payload.customer
  },

```
Denne oppdaterer staten som blir initialisert når storen bygges:
**store/index.js:**
```JavaScript
export const state = () => ({
  admin: false, // Sier om brukeren er admin eller ikke
  customer: false, // Sier ifra om brukeren er kunde eller ikke
  employee: false, // Sier om brukeren er ansatt eller ikke
  error: null, // Holder feilmeldingen vår
  loading: false, // Brukes ved logg inn i det vi begynner autentiseringen
  user: null // Holder brukeren
})

```

Disse attributtene kan vi så bruke ved hjelp av getters.

**pages/login.vue:**
```JavaScript
// Henter vaiabler fra storen
  computed: {
    user () {
      return this.$store.getters.user
    },
    error () {
      return this.$store.getters.error
    },
    loading () {
      return this.$store.getters.loading
    }
  },
```





## Database
Databasen som blir brukt er Googles Firestore. Dette er en rask NoSQL database. Les mer om databasen vår [her](https://gitlab.stud.idi.ntnu.no/programvareutvikling-v19/gruppe-30/wikis/about/firestore).
Spørringer mot databasen krever ikke innlogging hvis man kjører programmet lokalt, da access tokens settes opp når prosjektet initialiseres.
Ønsker man å administrere databasen via UI kreves en google bruker. Send mail til jespergp@stud.ntnu.no, så fikses tilgang :rocket:
UI kan så aksesseres via https://console.firebase.google.com/u/0/project/pu30-5b0f9/overview


## Tech/Rammeverk brukt:

### Bygd med:
* [JavaScript ES6](http://ecma-international.org/publications/standards/Ecma-006.htm)
* [Vue](https://vuejs.org)
* [Vuetify](https://vuetifyjs.com/en/getting-started/why-vuetify)
* [NPM](https://www.npmjs.com/)
* [Firestore](https://firebase.google.com/docs/firestore/)
* [Node](https://nodejs.org/en/)
* [Nuxt](https://nuxtjs.org/)
* [Cloud functions](https://cloud.google.com/functions/)
* [ChartJS](https://www.chartjs.org/)
* [Nodemailer](https://nodemailer.com/about/)
* [Jest](https://jestjs.io/)

## Features:
* Moderne nettside :rocket:
* Kunde kan legge inn, endre og slette reservasjoner
* Kunder kan opprette brukere
* Admin kan administrere oppsettet for restauranten
* Ansatte kan enkelt opprette, endre, slette og se reservasjoner, samt se statusen til restauranten til enhver tid
* Sjef kan se statistikk over besøkende for å se hvilke måneder, dager og tidspunkt som er mest attraktive og tilpasse driften etter dette

## Tester:
Vi benytter oss av Jest for rammeverk av testing.
For å kjøre testene kreves følgende (Hopp til steg 3 hvis du har jest):
1. `$ npm install -g jest`
2. `$ npm install -g vue-jest`
3. `$ npm test -- 'tests'`

Du får så tilbakemelding i konsollen om testene fungerte eller ei.
Nye tester kan enkelt legges til under test-mappen.

Eks på test:
```JavaScript
test('Updates table when setTable is called with table', () => {
  const store = new Vuex.Store(cloneDeep(Store))
  const testTable = {
    tableID: 1,
    capacity: 2,
    currently: 3,
    occupied: true
  }
  expect(store.getters.tables.length).toBe(0)
  store.commit('setTable', testTable)
  expect(store.state.tables[0]).toBe(testTable)
})
```

## Gitflow
#### Feature
Vi benytter oss av Gitflow som workflow. Alle nye features lages i en feature branch fra develop, som merges tilbake til develop når den er ferdig. Når vi har nok features lager vi en release branch
som går igjennom testing, før denne merges inn i master og tilbake til develop.

#### Hotfix
Hvis vi fikser bugs lages disse i hotfix branches ut fra hvilken branch man fikser, og merges tilbake til den respektive branchen. Unntak er hvis det er feil i master branchen, da må hotfixen merges inn i både master
og develop. 


## Versjoner:
Les mer om versjonene [her](CHANGELOG.md)

## Mappestruktur:
* Components (Inneholder komponenter som brukes flere steder eller har mulighet til å brukes flere steder)
  * AddReservation.vue (Brukes av booking.vue for å opprette en ny reservasjon av ansatte)
  * AddTable.vue (Brukes av tableEditor.vue for å legge til/slette eller endre bord i restauranten)
  * BarChart.js (Brukes av statistikksiden for å rendere stolpe diagrammet)
  * ChangeReservation.vue (Brukes av siden CustomerChangeReservation for å slette/endre en reservasjon)
  * CustomersReservations.vue (Brukes av dashboard.vue når kunder logger inn for å vise kundens reservasjoner)
  * NewReservation.vue (Brukes av customer-reservation.vue for å lage ny reservasjon)
  * TodaysTimelineForTable.vue (Brukes av ViewTable.vue for å vise kommende reservasjoner)
  * ViewTable.vue (Brukes av booking.vue for å opprette/vise informasjon om reservasjoner på valgt bord).
* Functions (Back-end funksjoner)
  * index.js (Inneholder alle cloud-funksjonene)
* Layouts (Inneholder de ulike layoutsene som brukes)
  * ansattLoggInn.vue (Brukes på ansatt.vue)
  * default.vue (Brukes når man har logget inn)
  * frontpage.vue (Brukes når man ikke er logget inn)
  * index.vue (Brukes av forsiden)
* middleware (Inneholder ulike routing regler for å sjekke at man er logget inn og har nødvendige rettigheter for å besøke siden.)
  * admin.js (Sjekker om man er admin)
  * customer.js (Sjekker om man er kunde)
  * employee.js (sjekker om man er ansatt)
  * router-check.js (Sjekker at man er logget inn) 
* Pages (Inneholder alle sidene til prosjektet): 
  * allreservations.vue (Alle kommende reservasjoner)
  * ansatt.vue (Innloggingside for ansatte)
  * booking.vue (Restaurantoversikt)
  * customer-reservation.vue
  * customerChangeRegistration.vue
  * customerRegistration.vue
  * dashboard.vue (Den første siden man møter etter innlogging)
  * forgot-password.vue
  * index.vue 
  * login.vue 
  * loyalty-program.vue 
  * menu.vue
  * statistics.vue
  * tableEditor.vue (Endre oppsett av restauranten)
  * termsAndConditions.vue
* plugis (Tillegg som settes opp i starten)
  * fireinit.js (Setter opp tilgang til databasen)
  * logincheck.js (Sjekker autoinnlogging)
  * vue-inject.js (Lager globale variabler som kan brukes)
  * vuetify.js (Setter opp vuetify)
* static
* store
  * index.js (Tilstandskontrolløren)
* tests
  * store.spec.js


Laget med :heart: av Jesper Paulsen, Maria Brenna, Mathilde Tillman Hegdal, Olav Falck-Pedersen, Una Førde Schjetne og Wictor Zhao

CHANGELOG.md
# Changelog
Alle store endringer på prosjektet vil bli dokumentert her,

# Versjon 1.3.0 28.03.2019

Inneholder brukerhistorier 14, 15, 16, 17, 18 og 19. (#6, #8, #11, #14, #114, #117)

* 14: Brukervennlig side for kunde (#6)
  * 14 - 01 Front end forside (#107)
  * 14 - 02 Front end menyside (#108)
  * 14 - 03 Flytte ansatt logge inn på egen side (#111)

* 15 - Kunde opprette bruker (#8)
  * 15 - 01 Kunde opprette bruker(#109)

* 16 - Kunde logg inn (#11)
  * 16 - 01 Kunde logget inn (#110)

* 17 - Kunde logge ut (#14)

* 18 - Kunde glemt passord (#114)

* 19 - Kunde lojalitetsprogram (#117)

## Hotfixes
* Bug: Lojalitetsprogram (#122)
* Bug: Statistikk for reservasjoner etter klokken 22 dukker ikke opp (#121)
* Kundereservasjon - Avbryt knapp ved bordbestilling (#113)
* Opprette/endre reservasjon: Sjekk på om bordet er ledig fungerer ikke (#104)
* Statistikk: Aksene begynner på 1 istedet for 0 (#104)
* Mail - endre starttid i mail (#97)
* Endre reservasjon - kan kun velge minutter 00, 15, 30, 45 (#95)
* Endre reservasjon - Burde ikke være mulig å reservere etter kl.22 (#94)
* Statistikk - Tidspunkt reservasjoner skal ha tidsintervall 12-24 (#93)
* Statistikk - Endre pr til pr. eller per i statistikk (#92)
* Kundereservasjon - Må velge gyldig antall personer ved bestilling fra kundens side

# v1.1.9 28.02.2019
Denne versjonen inneholder den ferdige brukerhistorien #7 (Issue #18):
07: Ansatte endre reservasjon
* Reservasjonsoversikt
* Endre reservasjon
* Slette reservasjon

### Hotfixes 
#### 01.03.2019:
* Alle reservasjoner viser nå kun reservasjoner fra og med starten av dagen og fram i tid
* Reservasjoner kan nå endre dato
* Allreservations sjekker nå om bordet er ledig før det kan bookes
* Bookingsystemet tillater ikke lenger flere gjester enn det er kapasitet til på bordet
* Fikset feil hvor ViewTable tillot booking tilbake i tid.
* FIkset feil hvor ViewTable ikke slo ut på alle tall i antall personer
#### 03.03.2019:
* Fikset feil som gjorde at functions mappen ikke ville bli deployet
#### 04.03.2019:
* Fikset feil som gjorde at det ikke alltid var allreservations oppdaterte til riktig tid ved enring. Fjernet counter som lå som hindring.

# v1.1.8 26.02.2019
Inneholder brukerhistoriene 4, 5, 6, 7 og 8 (Issues: #10, #15, #16, #18 og #19):
## Nytt i denne versjonen:
* 04: Ansatte kan logge inn
  * Store riktig state
  * Middleware på employee sider
* 05: Ansatte kan logge ut
* 06: Ansatte kan reservere bord:
  * Booking oversikt
  * Dialog for å opprette reservasjon
  * Dialog for å endre reservasjon
* 07: **WIP**: Ansatte endre reservasjon
  * Reservasjonsoversikt
  * Endre reservasjon
* 08: Ansatte slette reservasjon

# v1.1.4
Denne versjonen gjør det mulig for å ansatte og se statusen til restauranten, og de inviduelle bordene.
* Bookingsystem som viser grønn farge for ledige bord, og rød for opptatt.
* Info om de ulike bordene, og evt. om aktiv reservasjon
* Oppdatere pågående reservasjoner
* Opprette nye versjoner

# v1.1.0 21.02.2019
### Lagt til
* Nytt layout
* Ny forside
* Firebase
* Mulighet for å logge inn
* Mulighet for å logge ut
* Mulighet for å legge til bord
* Mulighet for å fjerne bord

### Hotfixes
* La til kommentarer
* Fikset problem hvor login knappen forsvant pga v-if="!user"
* Fikset problem hvor bord ble mutert i stedet for å bli kopiert i AddTable. Dette løste at man ikke lenger fikk feilmelding når man endret bordet, og at det ikke er mulig å opprette bord uten å oppgi bordnummer eller kapasitet.

# v1.0.9
Storen blir nå testet

# v1.0.8
Brukere med admin tilgang kan nå slette og legge til bord.
Bordene lastes og slettes via storen.

# v1.0.7 21.02.2019
* Mulighet for å logge ut

# v1.0.6 21.02.2019
Nuxt generate og ci/cd fungerer
### Lagt til
* Ryddet i mappestrukturen
* Bygget prosjektet fra bunnen av på nytt


functions/index.js
const  admin       = require('firebase-admin'),
      bodyParser  = require('body-parser'),
      cors        = require('cors'),
      express     = require('express'),
      functions   = require('firebase-functions'),
      moment      = require('moment'),
      request     = require('request'),
      // ses         = require('node-ses'),
      app         = express(),
      // Dato for åpningsdato
      openingDay = moment('2019-02-17', 'YYYY-MM-DD').valueOf(),
      // Dager siden åpningsdag
      days = moment().diff(openingDay, 'days'),
      // Uker siden åpningsdag
      weeks = moment().diff(openingDay, 'weeks'),
      nodemailer = require('nodemailer')


app.use(bodyParser.json())

// // Create and Deploy Your First Cloud Functions
// // https://firebase.google.com/docs/functions/write-firebase-functions
//
exports.helloWorld = functions.https.onRequest((request, response) => {
  let message = request.body.message
  response.send("Hello from Firebase! " + message)
})
// The Cloud Functions for Firebase SDK to create Cloud Functions and setup triggers.

// The Firebase Admin SDK to access the Firebase Realtime Database.

admin.initializeApp(functions.config().firebase);
let db = admin.firestore();
/*
Setter opp en mailtransporter med autentifisering gjennom gmail-kontoen til selskapet.
 */
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
let transporter = nodemailer.createTransport({
  service: 'gmail',
  auth: {
    user: 'tacosbetrippin@gmail.com',
    pass: 'trippin123'
  }
});

// 'use strict';

/*
Cloud-funksjon som henter informasjon om en nyopprettet reservasjon. Brukerens epost og navn, samt reservasjonsID og starttid
blir brukt til å generere en bekreftelsesepost til kunden.
 */
exports.sendWelcomeEmail = functions.https.onRequest((request, response) => {
  response.header('Access-Control-Allow-Origin', '*')
  response.header('Access-Control-Allow-Headers', 'Content-Type')
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  // ;
  let reservationID = request.body.reservationID; //henter brukerens reservasjonsID, epost, navn og tid for reservasjonen
  let email = request.body.email;
  let displayName = request.body.displayName;
  let startTime = request.body.startTime;
  console.log(email, displayName, reservationID, startTime);
  sendWelcomeEmail(email, displayName, reservationID, startTime)
    .then(response.send('Mail forsøkt sendt til ' + email))
    .catch(error => {
      console.log('Klarte ikke å sende epost')
      console.log(error)
      response.send(error)
    })
  }
);

exports.sendCancellationEmail = functions.https.onRequest((request, response) => {
    response.header('Access-Control-Allow-Origin', '*')
    response.header('Access-Control-Allow-Headers', 'Content-Type')
    response.set('Access-Control-Allow-Origin', "*")
    response.set('Access-Control-Allow-Methods', 'GET, POST')
    let reservationID = request.body.reservationID;
    let email = request.body.email;
    let displayName = request.body.displayName;
    let startTime = request.body.startTime;
    console.log(email);
    return sendCancellationEmail(email, displayName, reservationID, startTime)
      .then(response.send('Mail forsøkt sendt til ' + email))
      .catch(error => {
        console.log('Klarte ikke å sende epost')
        console.log(error)
        response.send(error)
      })
  }
);

const APP_NAME = 'Trippin Tacos';

/*
Funksjon som genererer og sender bekreftelsesepost til kunde. Formatterer tiden for reservasjonen, reservasjonsIDen og en
link til siden for å endre reservasjonen.
 */
function sendWelcomeEmail(email, displayName, reservationID, startTime) {
  console.log(email)
  const mailOptions = {
    from: `${APP_NAME} <noreply@firebase.com>`, //Avsender
    to: email, //Mottaker
  };
  const startDate = moment(Number(startTime)).format('DD.MM.YYYY');
  startTime = moment(Number(startTime)).add(2, 'hours').format('H:mm');
  mailOptions.subject = `Din reservasjon hos ${APP_NAME} - ${startDate}`;//Emne
  mailOptions.text = `Hei, ${displayName || ''}! \n\nDette er en bekreftelse på din reservasjon hos ${APP_NAME}.\nDu har reservert bord hos oss ${startDate || ''} klokken ${startTime || ''}.\nDitt reservasjonsnummer: ${reservationID || ''}\nØnsker du å endre reservasjonen, kan du følge denne linken: https://pu30-5b0f9.firebaseapp.com/customerChangeReservation\n\nVi ses!`;//Innmat
  return transporter.sendMail(mailOptions) //Sender eposten
    .then(() => {
      return console.log('New welcome email sent to:', email);
    })
    .catch(error => {
      console.log(error)
    })
}

function sendCancellationEmail(email, displayName, reservationID, startTime) {
  const mailOptions = {
    from: `${APP_NAME} <noreply@firebase.com>`,
    to: email,
  };
  const startDate = moment(Number(startTime)).format('DD.MM.YYYY');
  startTime = moment(Number(startTime)).add(2, 'hours').format('h:mm');
  mailOptions.subject = `Avbestilling av reservasjon hos ${APP_NAME}`;
  mailOptions.text = `Hei, ${displayName || ''}! \n\nDette er en bekreftelse på din avbestilling hos ${APP_NAME}.\nAvbestillingen gjelder følgende reservasjon:\n\n${startDate || ''} klokken ${startTime || ''}.\nReservasjonsnummer: ${reservationID || ''}\n\nØnsker du å lage en ny reservasjonen, kan du følge denne linken: https://pu30-5b0f9.firebaseapp.com/customer-reservation\n\nHåper vi ser deg en annen gang!`;
  return transporter.sendMail(mailOptions)
    .then( () => {
      return console.log('Cancellation email sent to:', email);
    })
    .catch(error => {
      console.log(error)
    })
}

/*
Cloud funksjon som regner ut hvor mange reservasjoner vi har hver hele time på restauranten. Statistikken runder ned til hver påbegynte klokketime, så reservasjoner
som er fra f.eks 12:00 til 13:59 vil regnes som ekstra besøk kl 12 og 13, men ikke 14.
 */
exports.hourlyNumberOfReservations = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let hourlyStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  console.log('/hourlyNumberOfReservations')
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const startTime = moment(reservation.startTime).format('H')
        const endTime = moment(reservation.endTime).format('H')
        for (let i = Number(startTime); i < Number(endTime) + 1; i++) {
          hourlyStatistics[i - 1] = hourlyStatistics[i - 1] + 1
        }
      })
    })
    .then(() => {
      let returnStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
      for (let i = 0; i < hourlyStatistics.length; i++) {
        if (i > 11 && i < 24) {
          returnStatistics[i - 12] = Number((hourlyStatistics[i - 1] / days).toFixed(5))
        }
      }
      return response.status(200).send(returnStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall reservasjoner')
      console.log(error)
      response.send(error)
    })
})

/*
Cloud funksjon som regner ut hvor mange besøkende vi har hver hele time på restauranten. Statistikken runder ned til hver påbegynte klokketime, så reservasjoner
som er fra f.eks 12:00 til 13:59 vil regnes som ekstra besøk kl 12 og 13, men ikke 14.
 */
exports.hourlyNumberOfPersons = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let hourlyStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  console.log('/hourlyNumberOfPersons')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const startTime = Number(moment(reservation.startTime).format('H'))
        const endTime = Number(moment(reservation.endTime).format('H'))
        for (let i = Number(startTime); i < Number(endTime) + 1; i++) {
          hourlyStatistics[i - 1] = hourlyStatistics[i - 1] + Number(reservation.numberOfPersons)
        }
      })
    })
    .then(() => {
      let returnStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
      for (let i = 0; i < hourlyStatistics.length; i++) {
        if (i > 10 && i < 25) {
          returnStatistics[i - 11] = Number((hourlyStatistics[i -1] / days).toFixed(5))
        }
      }
      return response.status(200).send(returnStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall besøkende')
      console.log(error)
      response.send(error)
    })
})
/*
Cloud funksjon som regner ut hvor mange reservasjoner vi har hver ukedag
 */
exports.dailyNumberOfReservations = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let dailyStatistics = [0, 0, 0, 0, 0, 0, 0]
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  console.log('/dailyNumberOfReservations')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const weekday = Number(moment(reservation.startTime).day())
        dailyStatistics[weekday] = dailyStatistics[weekday] + 1
      })
    })
    .then(() => {
      for (let i = 0; i < dailyStatistics.length; i++) {
        dailyStatistics[i] = dailyStatistics[i] / weeks
      }
      console.log(dailyStatistics)
      return response.status(200).send(dailyStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall besøkende')
      console.log(error)
      response.send(error)
    })
})

/*
Cloud funksjon som regner ut hvor mange besøkende vi har hver ukedag
 */
exports.dailyNumberOfPersons = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let dailyStatistics = [0, 0, 0, 0, 0, 0, 0]
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  console.log('/dailyNumberOfPersons')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const weekday = Number(moment(reservation.startTime).day())
        dailyStatistics[weekday] = dailyStatistics[weekday] + Number(reservation.numberOfPersons)
      })
    })
    .then(() => {
      for (let i = 0; i < dailyStatistics.length; i++) {
        dailyStatistics[i] = dailyStatistics[i] / weeks
      }
      console.log(dailyStatistics)
      return response.status(200).send(dailyStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall besøkende')
      console.log(error)
      response.send(error)
    })
})


/*
Cloud funksjon som regner ut hvor mange reservasjoner vi har hver måned
 */

exports.monthlyNumberOfReservations = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let monthlyStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  console.log('/monthlyNumberOfReservations')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const month = Number(moment(reservation.startTime).month())
        monthlyStatistics[month] = monthlyStatistics[month] + 1
      })
    })
    .then(() => {
      return response.status(200).send(monthlyStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall besøkende')
      console.log(error)
      response.send(error)
    })
})

/*
Cloud funksjon som regner ut hvor mange besøkende vi har hver måned
 */

exports.monthlyNumberOfPersons = functions.https.onRequest((request, response) => {
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  let monthlyStatistics = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  response.set('Access-Control-Allow-Origin', "*")
  response.set('Access-Control-Allow-Methods', 'GET, POST')
  console.log('/monthlyNumberOfPersons')
  db.collection('reservations')
    .get()
    .then(reservations => {
      return reservations.forEach(reservation => {
        reservation = reservation.data()
        const month = Number(moment(reservation.startTime).month())
        monthlyStatistics[month] = monthlyStatistics[month] + Number(reservation.numberOfPersons)
      })
    })
    .then(() => {
      return response.status(200).send(monthlyStatistics)
    })
    .catch(error => {
      console.log('Klarte ikke å mounte timestatistikk for antall besøkende')
      console.log(error)
      response.send(error)
    })
})




/*
Funksjon som oppdaterer alle feltene i reservasjonsbasen til true. Ble utført 11.03 før kundesystemet ble satt i prod,
ettersom alle reservasjoner fram til dette var gjort via systemet i restauranten.
 */
/*exports.databaseChangeDropin = functions.https.onRequest((request, response) => {
  console.log('/databaseChangeDropin')
  counter = 0
  db.collection('reservations').get()
    .then(reservations => {
      reservations.forEach(reservation => {
        reservation = reservation.data()
        reservation.dropIn = true
        db.collection('reservations').doc(reservation.reservationID + '').set(reservation)
          .then(() => {
            counter++
          })
          .catch(error => {
            console.log('Klarte ikke å oppdatere reservasjon med ID ' + reservation.reservationID)
            console.log(error)
          })
      })
    })
    .catch(error => {
      console.log('Klarte ikke å hente reservasjoner')
      console.log(error)
    })
  response.send('Oppdaterte ' + counter)
})*/
